src/coreclr/src/jit/lclmorph.cpp

```diff
@@ -521,6 +521,31 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                 PopValue();
                 break;
 
+            case GT_RETURN:
+                if (node->gtType != TYP_VOID)
+                {
+                    EscapeValue(TopValue(0), node);
+
+                    GenTree* retVal = node->AsUnOp()->gtOp1;
+                    if (retVal->OperIs(GT_LCL_VAR))
+                    {
+                        unsigned   lclNum = TopValue(0).LclNum();
+                        LclVarDsc* varDsc = m_compiler->lvaGetDesc(lclNum);
+                        if (varDsc->lvPromotedStruct() && (varDsc->lvFieldCnt != 1))
+                        {
+                            assert(m_compiler->compNoReturnRetyping() || varDsc->lvDoNotEnregister);
+                            // Its fields were accessed, so do not enregister it.
+                            // TODO seandree: it is possible to enregister that if we add a gather operation in codegen.
+                            // like struct s { bool a -> rax, bool b -> rdx, bool c -> rcx } -> rax,
+                            // return s -> gather(a, b, c) from their registers into rax.
```

Comment:
Yes, I will change the description.

> I would reword the above as "We can no longer promote or enregister this struct, since it is referenced as a whole." or something like that. It's not really that its fields are accessed.

I am not sure that we are talking about the same case. In the case that is covered by this code we have a struct that fits into a register, like `return struct A {bool a; bool b;}`, so we, in theory, can put it in a register. However, if we had accessed its fields, like `A.a = 1; A.b =2; ` then we can't put the whole struct in a register, we have to put it on the stack. The check for "Were the field accessed?" is implicit, if we had not accessed them we would not have promoted it.
And the special case covers structs with 1 field, in this case, we still can put it in a register.

Reply:
Right. We could *in theory* put it into a register, but we don't currently have any support for accessing a multi-field struct as a whole. So perhaps "We can not promote this struct since it is referenced as a whole, and we don't currently have support for fully enregistering structs."
FWIW, enregistering (entire) structs is the second objective mentioned here: https://github.com/dotnet/runtime/blob/master/docs/design/features/first-class-structs.md#objectives (though perhaps it should be worded more clearly)

Reply:
I've deleted the changes from lclmorph.cpp and moved that logic into `morph.cpp:fgMorphSmpOp:GT_RETURN` case, because there we have more information.